x/evm/genesis.go

```diff
@@ -39,30 +38,31 @@ func InitGenesis(
 	for _, account := range data.Accounts {
 		address := common.HexToAddress(account.Address)
 		accAddress := sdk.AccAddress(address.Bytes())
+
 		// check that the EVM balance the matches the account balance
 		acc := accountKeeper.GetAccount(ctx, accAddress)
 		if acc == nil {
 			panic(fmt.Errorf("account not found for address %s", account.Address))
 		}
 
-		ethAcct, ok := acc.(evmostypes.EthAccountI)
-		if !ok {
-			panic(
-				fmt.Errorf("account %s must be an EthAccount interface, got %T",
-					account.Address, acc,
-				),
-			)
-		}
 		code := common.Hex2Bytes(account.Code)
 		codeHash := crypto.Keccak256Hash(code)
 
-		// we ignore the empty Code hash checking, see ethermint PR#1234
-		if len(account.Code) != 0 && !bytes.Equal(ethAcct.GetCodeHash().Bytes(), codeHash.Bytes()) {
-			s := "the evm state code doesn't match with the codehash\n"
-			panic(fmt.Sprintf("%s account: %s , evm state codehash: %v, ethAccount codehash: %v, evm state code: %s\n",
-				s, account.Address, codeHash, ethAcct.GetCodeHash(), account.Code))
-		}
-
+		// TODO: I think this can be removed now that the EVM state is detached from the account keeper state
+		// This basically was cross-checking that both are in sync.
+		//
+		// // we ignore the empty Code hash checking, see ethermint PR#1234
+		// if len(account.Code) != 0 && !bytes.Equal(ethAcct.GetCodeHash().Bytes(), codeHash.Bytes()) {
+		// 	s := "the evm state code doesn't match with the codehash\n"
+		// 	panic(fmt.Sprintf("%s account: %s , evm state codehash: %v, ethAccount codehash: %v, evm state code: %s\n",
+		// 		s, account.Address, codeHash, ethAcct.GetCodeHash(), account.Code))
+		// }
+
+		// TODO: Do we need to add the code hash to the genesis accounts too?
+		//
+		// TODO: what is the significance of the code hash? Why do both need to be stored and not just
+		// the code related to the account?
+		k.SetCodeHash(ctx, address, codeHash)
```

Comment:
I was unsure of the significance of the code hash - from what I understand this is to save storage when multiple contracts have the exact same byte code? 

However, we have extra entries pointing to the hash for each smart contract address - is this separation of storing the code and the codehash really necessary or could this be unified? I'm wondering if this is actually saving space or is actually only redundant

Reply:
> is to save storage when multiple contracts have the exact same byte code? 

yes

> we have extra entries pointing to the hash for each smart contract address

exactly, it's like a junction table in SQL in a m-n relationship